sycl/doc/design/PropagateCompilerFlagsToRuntime.md

```diff
@@ -0,0 +1,66 @@
+
+```
+clang++ -c test_host.cpp -o test_host.o
+clang++ -c -fsycl test_device_1.cpp -o test_device_1.o
+clang++ -c -fsycl -O0 test_device_2.cpp -o test_device_2.o
+clang++ -fsycl -o test test_host.o test_device_1.o test_device_2.o
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test_device_2.c module, some of the debuggability is lost.
+
+Another scenario is shown below:
+
+```
+clang++ -c -O0 -fsycl -g test.cpp -o test.o
+clang++ -fsycl test.o -o test
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test.cpp module, some of the debuggability is lost. The user was not able to set a breakpoint inside device code.
+
+## Requirements
+
+In order to support module-level debuggability, the user will compile different module files with different levels of optimization. These optimization levels must be preserved and made use of during the backend compilation. Following are the requirements for this feature.
+- If the user specifies `-Ox` as a front-end compile option for a particular module, this option must be preserved during backend JIT compilation.
+- If the user specifies `-Ox` option using the environment variable, this option will override any front-end compile option and the new option will be preserved during JIT compilation.
+- If the user specifies `-O0` option, SYCL runtime needs to pass the appropriate backend option to JIT compilation stages.
+
+The following table specifies the appropriate backend options for level-zero and OpenCL backends.
+
+| Front-end option | L0 backend option | OpenCL backend option |
+| ---------------- | ----------------- | --------------------- |
+|      -O0         |  -ze-opt-disable  |   -cl-opt-disable     |
+|      -O1         |  -ze-opt-level=1  |   /* no option */     |
+|      -O2         |  -ze-opt-level=1  |   /* no option */     |
+|      -O3         |  -ze-opt-level=2  |   /* no option */     |
+
+
+## Proposed design
+
+This chapter discusses changes required in various stages of the compilation pipeline.
+
+
+### Changes to the DPC++ front-end
+
+For each SYCL kernel, we add a new function attribute that is named `sycl-optlevel`. Value of this attribute is set to the optimization level which was used to compile the overlying module.
+
+### Changes to the sycl-post-link tool
+
+During `sycl-post-link` stage, a set of optional kernel features are combined to form a hash value and this hash value is used as a key to split a module into multiple sub-modules. Current list of optional kernel features include: (1) SYCL aspects (2) `large-grf` mode (3) `reqd-work-group-size`. In this design, we add the optimization level associated with the kernel into this list. This helps us to split the kernels based on their optimization level.
```

Comment:
The exact mechanism of grouping kernels together (I mean, hash value as a key) may change in the future, as well as the list of optional kernel features. I think we can say something more generic here, like:

> During device code split stage, optimization level attribute `sycl-optlevel` is treated as an optional feature, i.e. device code split algorithm ensures that no kernels with different values of `sycl-optlevel` are bundled into the same device image. See also optional kernel features [design document](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#changes-to-the-post-link-tool).

Reply:
Modified this. Thanks